### PR TITLE
types: BIT to CHAR conversion should change to uint first

### LINKED ISSUE #31124: type conversion from `BIT` to `CHAR` should always consider uint
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
drop table if exists t;
create table t (a bit(45));
insert into t values (117471723421);
alter table t modify column a varchar(31) collate utf8mb4_general_ci;
select * from t;
```

### 2. What did you expect to see? (Required)

MySQL 8.0.27 and MySQL 5.7.36:
```sql
+--------------+
| a            |
+--------------+
| 117471723421 |
+--------------+
1 row in set (0.03 sec)
```

### 3. What did you see instead (Required)

```sql
+----------------+
| a              |
+----------------+
| 0x001B59DC3F9D |
+----------------+
1 row in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

f39a32815

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #31124

I try it on MySQL 5.7.36 and 8.0.27:
```sql
mysql> create table t (a bit(45));
Query OK, 0 rows affected (0.02 sec)

mysql> insert into t values (117471723421);
Query OK, 1 row affected (0.01 sec)

mysql> alter table t modify column a varchar(31) collate utf8mb4_general_ci;
Query OK, 1 row affected (0.05 sec)
Records: 1  Duplicates: 0  Warnings: 0

mysql> select * from t;
+--------------+
| a            |
+--------------+
| 117471723421 |
+--------------+
1 row in set (0.03 sec)
```

We can see that even if non-binary string, the `bit` type is changed to `int` first instead of `varchar` directly. This is conflict with https://github.com/pingcap/tidb/issues/25037#issuecomment-852985135:
> 2: when cast the bit to binary, we should consider to convert it to uint then cast uint to string, rather than takeing the bit to string directly.

### What is changed and how it works?

Always change to uint first.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: types/datum.go [hunk 1]
```diff
@@ -964,17 +964,13 @@ func (d *Datum) convertToString(sc *stmtctx.StatementContext, target *FieldType)
...
-		if target.Tp == mysql.TypeString || (target.Tp == mysql.TypeVarchar && target.Collate == charset.CollationBin) {
-			val, err := d.GetBinaryLiteral().ToInt(sc)
-			if err != nil {
-				s = d.GetBinaryLiteral().ToString()
-			} else {
-				s = strconv.FormatUint(val, 10)
-			}
-		} else {
+		// https://github.com/pingcap/tidb/issues/25037.
+		// Consider converting to uint first.
```
Comment (AilinKid) [inline types/datum.go:968]: better add this to here 'https://github.com/pingcap/tidb/pull/31031#issuecomment-1001404832.'

**Review by AilinKid — APPROVED** [review]
Rest LGTM

**Review by xiongjiwei — APPROVED** [review]

**Pull request merged**